### PR TITLE
fix/nae-115/failure action response

### LINKED ISSUE #115: CreateFailure action response
When a `CreateFailure` Action is dispatched, there's no entity information attached. If possible that information could be useful to display to the user. Not only that it failed, but what entity failed. Especially for historical purposes if you're tracking a user events.

### PR DESCRIPTION
**Pull request by schuchard**
- provide entity/entities to *failure actions
- provide key(s) to DeleteByKeys and DeleteManyByKeys
- update specs
- typing refactors and improvements
- narrows type on reducer from string to `EntityActionTypes`, fix lint error in reducer
- bump lib version

fixes #115

### CODE REVIEW

File: projects/ngrx-auto-entity/src/lib/decorators/entity.ts [hunk 1]
```diff
@@ -12,20 +12,23 @@ export interface IEffectExcept {
...
  */
-export interface IEntityTransformer {
-  fromServer?: (data: any, criteria?: any) => any;
-  toServer?: (entity: any, criteria?: any) => any;
+export interface IEntityTransformer<T = any, U = any, V = any> {
+  fromServer?: (data: U, criteria?: any) => T;
+  toServer?: (entity: T, criteria?: any) => V;
+}
+
+export interface IEntityInfoBase {
```
Comment (jrista) [inline projects/ngrx-auto-entity/src/lib/decorators/entity.ts:20]: Lets call this `IEntityNames`.

**Review by jrista — CHANGES_REQUESTED** [review]

**Comment by jrista** [comment]
Lets squash all the commits for this into one, with a proper commit message as described last night:

```
feat|fix|refactor: <message>

 + Additions
 - Removals
 * Other stuff

<related issues>
```

**Comment by schuchard** [comment]
ready to go @jrista

**Comment by schuchard** [comment]
What do you think about extending this pattern to the `Load*` actions and expose the following?

- `LoadFailure` - `key`
- `LoadManyFailure`, `LoadAllFailure` - `criteria`
- `LoadPageFailure` - `page`
- `LoadRangeFailure` - `range`

Another question, should also be passing the criteria in the failure cases as well. There could be identifying information in there as well, but this could be in a later PR?

**Comment by schuchard** [comment]
Just realized the types are not accurate for the action. `entity`, `entities`, `key`, and `keys` are not coming through no the action type. Need to make a few more changes.

**Comment by schuchard** [comment]
false alarm, it turns out we need to explicitly type the action to pickup the properties.

```ts
ofEntityType<Customer, CreateFailure<Customer>>
```

**Comment by jrista** [comment]
> What do you think about extending this pattern to the `Load*` actions and expose the following?
> 
> * `LoadFailure` - `key`
> * `LoadManyFailure`, `LoadAllFailure` - `criteria`
> * `LoadPageFailure` - `page`
> * `LoadRangeFailure` - `range`
> 
> Another question, should we also be passing the criteria in the failure cases as well? There could be identifying information in there as well, but this could be in a later PR?

Yeah, there is no reason we can't thread all the relevant information through the entire chain of actions.

We might as well go ahead and do that. One of the other stories is also to expand our action patterns from a two-stage to a three-stage patern. From Initiaor -> Result to Initiator -> Response -> Result. I guess it is up to you whether you want to make the changes you recommended here, or maybe wait for the three-stage pattern story, and make them when we make that change?

**Review by jrista — DISMISSED** [review]

File: projects/ngrx-auto-entity/src/lib/decorators/entity.ts (hunk 1 above)
Comment (schuchard) [reply projects/ngrx-auto-entity/src/lib/decorators/entity.ts]: updated

**Review by jrista — APPROVED** [review]

**Pull request merged**